[sparse] implementing transposition rule for `sparse.coo_matmat`.

dimasad opened the pull request:
The transposition rule for `sparse.coo_matmat` is very straightforward, and the code is very similar to `_coo_matvec_transpose`, already implemented in `jax.experimental.sparse.ops.py`.

dimasad committed:
ENH: implementing transposition rule for `sparse.coo_matmat`.

dimasad commented:
@googlebot I signed it!

jakevdp commented:
Thanks for this! Just FYI, the transpose rule is a different sense of the word "transpose" than in the issue in #7533: transpose rules are used to implement reverse-mode autodiff (vjp). The reason I stopped before implementing it is because it became clear that batching was more important (e.g. you can't use `jacfwd` or `jacrev` to test autodiff without batching being implemented) so I changed my focus to developing the batched COO format (BCOO in that file) which now has batching & full autodiff implemented.

If you're interested in contributing this transpose rule, we'd also need to add a test that exercises it similar to coo_matvec's autodiff test here: https://github.com/google/jax/blob/38cb2b530aaecc74443de373f27a0ee143b6166d/tests/sparse_test.py#L319-L363

dimasad commented:
> Thanks for this! Just FYI, the transpose rule is a different sense of the word "transpose" than in the issue in #7533: transpose rules are used to implement reverse-mode autodiff (vjp). 

Yes, I'm aware, of this, I implemented this code because I was using `jax.grad` to perform reverse differentiation of a cost function that used a matrix-matrix multiplication like in #7533. I was able to work around the bug by simply transposing the whole expression twice, but for reverse differentiation I used this code I just contributed. I opened the pull request because I think it might be useful to more people.

I'll implement a few tests like the `matvec` counterparts, perhaps Monday.

dimasad committed:
Added test case for AD of `coo_matmat`.

dimasad committed:
Fixed AD tests of `coo_matmat`.

dimasad commented:
> If you're interested in contributing this transpose rule, we'd also need to add a test that exercises it similar to coo_matvec's autodiff test here:
> 
> https://github.com/google/jax/blob/38cb2b530aaecc74443de373f27a0ee143b6166d/tests/sparse_test.py#L319-L363
> 
> The "reverse mode" sections there are where the transpose rule will be called.

Added test coverage of the implemented code with minimal changes to the existing code. By what was written in the TODO comments I just updated the `matvec_ad` test to test `matmul` in a way that covers both `coo_matvec` and `coo_matmat`. If separate tests are wanted I can just change it. All updated tests passed here.

jakevdp commented:
Looking good - you can address the test failures by modifying the tolerances.

Once you have that fixed, please squash your changes into a single commit (see https://jax.readthedocs.io/en/latest/contributing.html#single-change-commits-and-pull-requests)

Then we can get this in!

jakevdp commented:
It looks like `coo_matvec_transpose` has been added in the meantime. Sorry we weren't able to merge this contribution!
